### PR TITLE
ParenthesisNode and parenthesis configuration options for v2

### PR DESCRIPTION
**Pull request by FSMaxB**
It's finally ready, including an update of the pretty printing example page and an update of the documentation.

This introduces a new `ParenthesisNode` that will always get parsed into the node tree. The `toString` and `toTex` functions then use or ignore those `ParenthesisNodes` based on the parenthesis configuration.

The new parenthesis configuration option has the following three options:
* `keep`: Print parentheses in the same way as they were in the input (default)
* `auto`: Sanitize parentheses. This is the way it has been in v1.
* `all`: Print all parentheses to make the structure of the node tree explicit (this also ignores `ParenthesisNodes`).

What's still missing is a way to pass the configuration to `toTex` and `toString` directly, but I want to have this code reviewed first before implementing that.

How I want to implement the missing part:
* Every function can have a `.toString` or `.toTex` property that is either a callback function or a template string.
* Both `toTex` and `toString` can have two parameters
  - `config`: A config object that locally overrides the global parenthesis option
  - `callback`: The same callback as right now. (this is still necessary because the `.toString` and `.toTex` properties aren't powerful enough to handle all the cases that those callbacks can [like overwriting `toTex` and `toString` of the nodes for example] and it allows changing the behaviour for a single function call). This callback will override the `.toString` and `.toTex` properties.

I'm asking everyone to play around with the LaTeX output and the different parenthesis options (via `examples/browser/pretty_printing_with_mathjax.html`) because this topic is quite complex and there can very well be bugs that slipped through ( even ones that are still present in v1). But if you do so, don't forget to run `npm run build` beforehand. One thing to notice though: The LaTeX output for `FunctionNode`s is still pretty ugly because it isn't able to sanitize parentheses, so there are too many of them in most cases. I haven't tackled this problem so far.

### CODE REVIEW

**Comment by FSMaxB** [comment]
I changed one of the commit messages (from 'auto' to 'all')

**Comment by josdejong** [comment]
Thanks, I will check it out asap.

File: lib/expression/node/ParenthesisNode.js [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+  /**
+   * Get string representation
+   * @return {String} str
+   * @override
+   */
+  ParenthesisNode.prototype._toString = function() {
+    if (config.parenthesis === 'keep') {
+      return '(' + this.content.toString() + ')';
+    }
+    return this.content._toString();
```
Comment (josdejong) [inline lib/expression/node/ParenthesisNode.js]: Shouldn't 
```js
return this.content._toString();
```
be 
```js
return this.content.toString();`
```
?

**Comment by josdejong** [comment]
Looks good, thanks! This is a nice new configuration option. The code for determining whether or not to use parenthesis is quite complex here and there, but you coded it very readable.

**Comment by FSMaxB** [comment]
Yes, you're right, this shouldn't contain an underscore, I'll fix that. I'll have another pull request ready soon. That one will be the last API breaking change, everything after that will only extend the API.

**Pull request merged**